Core/Core/Files/FileUploader.swift

```diff
@@ -0,0 +1,325 @@
+        Logger.shared.log()
+        performAndWait { context in
+            guard let file: File = context.first(where: #keyPath(File.taskIDRaw), equals: NSNumber(value: dataTask.taskIdentifier)) else {
+                return
+            }
+            do {
+                let decoder = JSONDecoder()
+                decoder.dateDecodingStrategy = .iso8601
+                let response = try decoder.decode(APIFile.self, from: data)
+                Logger.shared.log("Created a file with id \(response.id)")
+                try file.update(fromAPIModel: response)
+            } catch {
+                file.uploadError = error.localizedDescription
+                Logger.shared.error(error)
+            }
+            do {
+                try context.save()
+            } catch {
+                Logger.shared.error(error)
+            }
+        }
+    }
+
+    public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        Logger.shared.log()
+        if let error = error {
+            Logger.shared.error(error)
+        }
+        performAndWait { context in
+            guard let file: File = context.first(where: #keyPath(File.taskIDRaw), equals: NSNumber(value: task.taskIdentifier)) else {
+                return
+            }
+            file.uploadError = error?.localizedDescription ?? file.uploadError
+            file.taskID = nil
+            do {
+                try context.save()
+            } catch {
+                Logger.shared.error(error)
+            }
+            if let assignmentID = file.assignmentID, let courseID = file.courseID {
+                self.submit(assignmentID: assignmentID, courseID: courseID)
+            }
+        }
+    }
+
+    public func urlSessionDidFinishEvents(forBackgroundURLSession session: URLSession) {
+        Logger.shared.log()
+        self.completionHandler?()
+        session.finishTasksAndInvalidate()
+    }
+
+    private func getTarget(context: FileUploadContext, name: String, callback: @escaping (FileUploadTarget?, Error?) -> Void) {
+        Logger.shared.log()
+        let body = PostFileUploadTargetRequest.Body(name: name, on_duplicate: .rename, parent_folder_id: nil)
+        let request = PostFileUploadTargetRequest(context: context, body: body)
+        api.makeRequest(request) { target, _, error in
+            callback(target, error)
+        }
+    }
+
+    private func performAndWait(block: (NSManagedObjectContext) -> Void) {
+        let context = database.newBackgroundContext()
+        context.performAndWait {
+            block(context)
+        }
+    }
+}
+
+// MARK: - Submissions
+extension FileUploader {
+    private func submit(assignmentID: String, courseID: String) {
+        performAndWait { context in
+            let files: [File] = context.all(where: #keyPath(File.assignmentID), equals: assignmentID)
+            let ready = files.allSatisfy { $0.isUploaded } == true
```

Comment:
is the `== true` redundant here? `allSatisfy` returns a `Bool`.

Reply:
Ah yes. This used to be optional. Fixed.